packages/core/src/hydration/error_handling.ts

```diff
@@ -6,39 +6,366 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import {TNode} from '../render3/interfaces/node';
+import {RuntimeError, RuntimeErrorCode} from '../errors';
+import {getDeclarationComponentDef} from '../render3/instructions/element_validation';
+import {TNode, TNodeType} from '../render3/interfaces/node';
 import {RNode} from '../render3/interfaces/renderer_dom';
-import {LView} from '../render3/interfaces/view';
+import {LView, TVIEW} from '../render3/interfaces/view';
+import {getParentRElement} from '../render3/node_manipulation';
+
+import {NodeNavigationStep} from './interfaces';
+
+
+const AT_THIS_LOCATION = '<-- AT THIS LOCATION';
+
+function getFriendlyStringFromTNodeType(key: number): string {
```

Comment:
nit: I'd also propose to add a quick JSDoc comment to explain what this function does.
```suggestion
function getFriendlyStringFromTNodeType(tNodeType: TNodeType): string {
```